bundles/tools.vitruv.dsls.reactions/src/tools/vitruv/dsls/reactions/codegen/classgenerators/routine/CreateBlockClassGenerator.xtend

```diff
@@ -0,0 +1,123 @@
+	 * the defined elements.
+	 * 
+	 * @param typesBuilderExtensionProvider the Xtext types builder, must not be {@code null}
+	 * @param qualifiedClassName the qualified name of the class to create, may not be {@code null} or empty
+	 * @param createBlock the code block with create statements to create a class for, must not be {@code null}
+	 */
+	new(TypesBuilderExtensionProvider typesBuilderExtensionProvider, String qualifiedClassName,
+		CreateBlock createBlock) {
+		super(typesBuilderExtensionProvider)
+		checkArgument(!qualifiedClassName.nullOrEmpty, "class name must not be null or empty")
+		this.qualifiedClassName = qualifiedClassName
+		this.createBlock = checkNotNull(createBlock, "create block must not be null")
+		this.createdElementsClass = generateNewlyAccessibleElementsContainerClass(createdValuesClassQualifiedName)
+	}
+
+	private def getCreatedValuesClassQualifiedName() {
+		qualifiedClassName + "." + CREATED_ELEMENTS_SIMPLE_CLASS_NAME
+	}
+
+	override Iterable<AccessibleElement> getNewlyAccessibleElementsAfterExecution() {
+		return createBlock.createStatements.mapFixed [
+			new AccessibleElement(name ?: MISSING_NAME, metaclass?.javaClassName)
+		]
+	}
+
+	override getNewlyAccessibleElementsContainerType() {
+		return createdElementsClass
+	}
+
+	override generateEmptyClass() {
+		this.generatedClass = createBlock.toClass(qualifiedClassName) [
+			visibility = JvmVisibility.PRIVATE
+			static = true
+		]
+		return generatedClass
+	}
+
+	override generateBody() {
+		generatedClass => [
+			superTypes += typeRef(AbstractRepairRoutineRealization.Create)
+			members += createdElementsClass
+			members += generateConstructor()
+			members += generateMethodCreate()
+		]
+	}
+
+	def private generateConstructor() {
+		createBlock.toConstructor [
+			val reactionExecutionStateParameter = generateReactionExecutionStateParameter()
+			parameters += reactionExecutionStateParameter
+			body = '''super(«reactionExecutionStateParameter.name»);'''
+		]
+	}
+
+	private def JvmOperation generateMethodCreate() {
+		createBlock.toMethod(CREATE_ELEMENTS_METHOD_NAME, typeRef(newlyAccessibleElementsContainerType)) [
+			body = '''
+				«FOR createStatement : createBlock.createStatements»
+					«createStatement.elementCreationCode»
+				«ENDFOR»
+				return new «newlyAccessibleElementsContainerType»(« //
+				FOR createdElement : newlyAccessibleElementsAfterExecution SEPARATOR ", "»«createdElement.name»«ENDFOR»);
+			'''
+		]
+	}
+
+	private def StringConcatenationClient getElementCreationCode(NamedMetaclassReference elementCreate) {
+		val affectedElementClass = elementCreate.metaclass
+		val createdClassFactory = affectedElementClass?.EPackage?.EFactoryInstance?.runtimeClassName
+		return '''
+		«affectedElementClass.javaClassName» «elementCreate.name ?: MISSING_NAME» = «createdClassFactory?: 
+			MISSING_TYPE».eINSTANCE.create«affectedElementClass?.name?: MISSING_TYPE»();
+		notifyObjectCreated(«elementCreate.name ?: MISSING_NAME»);'''
+	}
```

Comment:
I must admit that I don't get all the language generation code but this seems to allow that either the created element name or type are not set? In both cases this would generate invalid Java source code I suppose. Can't we require values for these in the grammar?

Reply:
We could also abort code generation, but then content assist does not work as it depends on the generated code, which is why Xtext code generation should be able to deal with incomplete models. The grammar does also require values for these attribute (as you propose), which is why the language also indicates an error in such a case, but still code generation should be performed on such an incomplete model.

Reply:
Ah, thanks for clearing that up.